src/contractor/graph_contractor.cpp

```diff
@@ -568,28 +547,216 @@ bool GraphContractor::IsNodeIndependent(const std::vector<float> &priorities,
+
+    const NodeID number_of_nodes = graph.GetNumberOfNodes();
+
+    ThreadDataContainer thread_data_list(number_of_nodes);
+
+    NodeID number_of_contracted_nodes = 0;
+    std::vector<NodeID> new_to_old_node_id(number_of_nodes);
+    // Fill the map with an identiy mapping
+    std::iota(new_to_old_node_id.begin(), new_to_old_node_id.end(), 0);
+
+    ContractorNodeData node_data{graph.GetNumberOfNodes(),
+                                 std::move(node_is_uncontracted_),
+                                 std::move(node_is_contractable_),
+                                 std::move(node_weights_)};
+
+    std::vector<RemainingNodeData> remaining_nodes;
+    remaining_nodes.reserve(number_of_nodes);
+    for (auto node : util::irange<NodeID>(0, number_of_nodes))
     {
-        return hasha < hashb;
+        if (node_data.is_core[node])
+        {
+            if (node_data.contractable[node])
+            {
+                remaining_nodes.emplace_back(node, false);
+            }
+            else
+            {
+                node_data.priorities[node] =
+                    std::numeric_limits<ContractorNodeData::NodePriority>::max();
+            }
+        }
+        else
+        {
+            node_data.priorities[node] = 0;
+        }
     }
-    return a < b;
+
+    {
+        util::UnbufferedLog log;
+        log << "initializing node priorities...";
+        tbb::parallel_for(tbb::blocked_range<std::size_t>(0, remaining_nodes.size(), PQGrainSize),
+                          [&](const auto &range) {
+                              ContractorThreadData *data = thread_data_list.GetThreadData();
+                              for (auto x = range.begin(), end = range.end(); x != end; ++x)
+                              {
+                                  auto node = remaining_nodes[x].id;
+                                  BOOST_ASSERT(node_data.contractable[node]);
+                                  node_data.priorities[node] = EvaluateNodePriority(
+                                      SimulateNodeContraction(data, graph, node, node_data.weights),
+                                      node_data.depths[node]);
+                              }
+                          });
+        log << " ok.";
+    }
+
+    auto number_of_core_nodes = std::max<std::size_t>(0, (1 - core_factor) * number_of_nodes);
+    auto number_of_nodes_to_contract = remaining_nodes.size() - number_of_core_nodes;
+    util::Log() << "preprocessing " << number_of_nodes_to_contract << " ("
+                << (number_of_nodes_to_contract / (float)number_of_nodes * 100.) << "%) nodes...";
+
+    util::UnbufferedLog log;
+    util::Percent p(log, remaining_nodes.size());
+
+    const util::XORFastHash<> hash;
+
+    unsigned current_level = 0;
+    std::size_t next_renumbering = number_of_nodes * 0.35;
+    while (remaining_nodes.size() > number_of_core_nodes)
+    {
+        if (remaining_nodes.size() < next_renumbering)
```

Comment:
I have bisected failing test on windows and the different ordering comes from this line in 883a8ca501e0359da90e7917346320764de54179 
```
if (remaining_nodes.size() <= next_renumbering)
```
will make reordering at the same time as before because `<` is not equivalent to `>` after integer truncation

Reply:
Hrm interesting. That line only governs when we do a cache-optimizing renumbering. I think this might create an edge case for nodes with same priority. For those nodes we use a bias function https://github.com/Project-OSRM/osrm-backend/pull/4427/files#diff-e32f0fbbf989930fbb6323b16171f940R124 that decides which one wins the tie. The bias function depends on the node id. If we renumber that tie might change since the node ID is different.

This bug was present in the [old code](https://github.com/Project-OSRM/osrm-backend/blob/master/src/contractor/graph_contractor.cpp#L116) as well. This should be solvable by always using the original node ID for the bias function.

Reply:
what wonders me is that it appears only on :windows: 
:+1: on proper tie breaking

Reply:
Hrm okay changing proper tie breaking doesn't seem to make a difference. This bug also affects OSX. I will try to compile OSRM on Windows, it seems to be reproducible in a consistent manner there?